File: packages/go/batches.go
package opensms

import (
	"bytes"
	"context"
	"mime/multipart"
	"net/http"
	"strconv"
)

// Batches is the batches resource, reached as client.Batches. A batch is
// created in the ready state and sends nothing until Start.
type Batches struct {
	http *transport
}

// Create creates a batch from JSON items (POST /v1/messages/batch). Invalid
// rows do not fail the request; they are counted and listed by Validation.
func (r *Batches) Create(ctx context.Context, params CreateBatchParams, opts ...CallOption) (*Batch, error) {
	var out Batch
	err := r.http.do(ctx, request{method: http.MethodPost, path: "/v1/messages/batch", body: params, idempotent: true, opts: applyCallOptions(opts)}, &out)
	if err != nil {
		return nil, err
	}
	return &out, nil
}

// CreateFromCSV creates a batch from CSV text with a to,text[,sender_id,...]
// header row (POST /v1/messages/batch, Content-Type text/csv). When
// params.Dedupe is set, the CSV is sent as a multipart upload with a dedupe
// field, the only form in which the API accepts that flag for CSV.
func (r *Batches) CreateFromCSV(ctx context.Context, csv []byte, params CreateBatchFromCSVParams, opts ...CallOption) (*Batch, error) {
	req := request{method: http.MethodPost, path: "/v1/messages/batch", rawBody: csv, contentType: "text/csv", idempotent: true, opts: applyCallOptions(opts)}
	if req.rawBody == nil {
		req.rawBody = []byte{}
	}
	if params.Dedupe != nil {
		var buf bytes.Buffer
		mw := multipart.NewWriter(&buf)
		part, err := mw.CreateFormFile("file", "batch.csv")
		if err == nil {
			_, err = part.Write(csv)
		}
		if err == nil {
			err = mw.WriteField("dedupe", strconv.FormatBool(*params.Dedupe))
		}
		if err == nil {
			err = mw.Close()
		}
		if err != nil {
			return nil, newLocalError("", "failed to build multipart body: "+err.Error(), err)
		}
		req.rawBody = buf.Bytes()
		req.contentType = mw.FormDataContentType()
	}
	var out Batch
	if err := r.http.do(ctx, req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Get fetches a batch (GET /v1/batches/{id}).
func (r *Batches) Get(ctx context.Context, id string) (*Batch, error) {
	if err := requireID("id", id); err != nil {
		return nil, err
	}
	var out Batch
	if err := r.http.do(ctx, request{method: http.MethodGet, path: "/v1/batches/" + escape(id)}, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Validation returns the per-row validation report
// (GET /v1/batches/{id}/validation).
func (r *Batches) Validation(ctx context.Context, id string) (*ValidationReport, error) {
	if err := requireID("id", id); err != nil {
		return nil, err
	}
	var out ValidationReport
	if err := r.http.do(ctx, request{method: http.MethodGet, path: "/v1/batches/" + escape(id) + "/validation"}, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Start starts a ready batch (POST /v1/batches/{id}/start).
func (r *Batches) Start(ctx context.Context, id string, opts ...CallOption) (*Batch, error) {
	if err := requireID("id", id); err != nil {
		return nil, err
	}
	var out Batch
	err := r.http.do(ctx, request{method: http.MethodPost, path: "/v1/batches/" + escape(id) + "/start", idempotent: true, opts: applyCallOptions(opts)}, &out)
	if err != nil {
		return nil, err
	}
	return &out, nil
}

// Stop stops a batch and cancels its unsent items
// (POST /v1/batches/{id}/stop).
func (r *Batches) Stop(ctx context.Context, id string, opts ...CallOption) (*BatchStopResult, error) {
	if err := requireID("id", id); err != nil {
		return nil, err
	}
	var out BatchStopResult
	err := r.http.do(ctx, request{method: http.MethodPost, path: "/v1/batches/" + escape(id) + "/stop", idempotent: true, opts: applyCallOptions(opts)}, &out)
	if err != nil {
		return nil, err
	}
	return &out, nil
}

// ListItems returns a page of the messages a batch created
// (GET /v1/batches/{id}/items).
func (r *Batches) ListItems(ctx context.Context, id string, params ListBatchItemsParams) (*Page[BatchItem], error) {
	if err := requireID("id", id); err != nil {
		return nil, err
	}
	q := newQuery().str("status", params.Status).int("limit", int64(params.Limit)).str("cursor", params.Cursor).values()
	var out Page[BatchItem]
	if err := r.http.do(ctx, request{method: http.MethodGet, path: "/v1/batches/" + escape(id) + "/items", query: q}, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

File: packages/php/src/Resources/Batches.php
<?php

declare(strict_types=1);

namespace Opensms\Resources;

use Opensms\Models;
use Opensms\Page;
use Opensms\Transport;

/**
 * The `batches` resource: bulk sends created in `ready` and started explicitly.
 * Accessed as `$client->batches`.
 */
final class Batches
{
    /** @internal */
    public function __construct(private readonly Transport $transport)
    {
    }

    /**
     * Create a batch from items. Params: `items` (list of
     * `{to, text, sender_id?, traffic_type?, callback_url?, metadata?}`),
     * `dedupe` (bool, default true). Returns the Batch (202, status ready).
     *
     * @param array{items: list<array<string, mixed>>, dedupe?: bool} $params
     * @param array{idempotencyKey?: string} $options
     * @return array<string, mixed>
     */
    public function create(array $params, array $options = []): array
    {
        $body = Models::map($params, ['items' => Models::SCALAR, 'dedupe' => Models::SCALAR], ['items']);
        if (!is_array($body['items'])) {
            throw new \InvalidArgumentException('OpenSMS: `items` must be a list.');
        }
        $body['items'] = array_map(
            static fn (array $item): array => Models::map($item, Models::BATCH_ITEM),
            array_values($body['items']),
        );

        return Models::object($this->transport->request('POST', '/v1/messages/batch', [], $body, [
            'idempotent' => true,
            'idempotencyKey' => Models::idempotencyKey($options),
        ]));
    }

    /**
     * Create a batch from CSV text (header row `to,text[,sender_id,...]`),
     * sent as `text/csv`. The server always dedupes a raw CSV body, so
     * `dedupe: false` switches the upload to multipart/form-data, the only
     * CSV form in which the API honours that flag.
     *
     * @param array{dedupe?: bool} $params
     * @param array{idempotencyKey?: string} $options
     * @return array<string, mixed>
     */
    public function createFromCsv(string $csv, array $params = [], array $options = []): array
    {
        $mapped = Models::map($params, ['dedupe' => Models::SCALAR]);
        $body = $csv;
        $contentType = 'text/csv';
        if (($mapped['dedupe'] ?? true) === false) {
            $boundary = 'opensms' . bin2hex(random_bytes(12));
            $body = "--{$boundary}\r\n"
                . "Content-Disposition: form-data; name=\"dedupe\"\r\n\r\n"
                . "false\r\n"
                . "--{$boundary}\r\n"
                . "Content-Disposition: form-data; name=\"file\"; filename=\"batch.csv\"\r\n"
                . "Content-Type: text/csv\r\n\r\n"
                . $csv . "\r\n"
                . "--{$boundary}--\r\n";
            $contentType = 'multipart/form-data; boundary=' . $boundary;
        }

        return Models::object($this->transport->request('POST', '/v1/messages/batch', [], null, [
            'idempotent' => true,
            'idempotencyKey' => Models::idempotencyKey($options),
            'rawBody' => $body,
            'contentType' => $contentType,
        ]));
    }

    /**
     * @return array<string, mixed>
     */
    public function get(string $id): array
    {
        return Models::object($this->transport->request('GET', '/v1/batches/' . Models::pathParam($id)));
    }

    /**
     * Per-row validation report: `{rows, total, valid, invalid, duplicates, suppressed}`.
     *
     * @return array<string, mixed>
     */
    public function validation(string $id): array
    {
        return Models::object($this->transport->request('GET', '/v1/batches/' . Models::pathParam($id) . '/validation'));
    }

    /**
     * Start sending a ready batch.
     *
     * @param array{idempotencyKey?: string} $options
     * @return array<string, mixed>
     */
    public function start(string $id, array $options = []): array
    {
        return Models::object($this->transport->request('POST', '/v1/batches/' . Models::pathParam($id) . '/start', [], null, [
            'idempotent' => true,
            'idempotencyKey' => Models::idempotencyKey($options),
        ]));
    }

    /**
     * Stop a batch. Returns `{id, status: "stopped", cancelled}`.
     *
     * @param array{idempotencyKey?: string} $options
     * @return array<string, mixed>
     */
    public function stop(string $id, array $options = []): array
    {
        return Models::object($this->transport->request('POST', '/v1/batches/' . Models::pathParam($id) . '/stop', [], null, [
            'idempotent' => true,
            'idempotencyKey' => Models::idempotencyKey($options),
        ]));
    }

    /**
     * List the messages a batch produced. Query: `status`, `limit`, `cursor`.
     *
     * @param array<string, mixed> $params
     */
    public function listItems(string $id, array $params = []): Page
    {
        $query = Models::map($params, Models::BATCH_ITEMS_QUERY, [], true);

        return Models::page($this->transport->request('GET', '/v1/batches/' . Models::pathParam($id) . '/items', $query));
    }
}
